### PR TITLE
Fixed the initialization

### LINKED ISSUE #200: Parameter not used for initialization
## Expected behavior
File:  hivemq-community-edition/src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java 

Function:
`public SubscriberWithIdentifiers(final @NotNull String subscriber, final int qos, final byte flags, final @Nullable String sharedName, final @NotNull ImmutableList<Integer> subscriptionIdentifier, final @Nullable String topicFilter)`

Parameter `subscriptionIdentifier` is not used for setting the `subscriptionIdentifier`

## Actual behavior
Creates a new empty list for `subscriptionIdentifier`

## Details
File link:
https://github.com/hivemq/hivemq-community-edition/blob/master/src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java#L52

The bug is still present in **master** branch.

- Affected HiveMQ CE version(s): 2020.4
- Used JVM version:

### PR DESCRIPTION
**Pull request by sakib-malek**
**Motivation**

Resolves #200 

**Changes**
Fixed the initialization of the `subscriptionIdentifier`

### CODE REVIEW

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public SubscriberWithIdentifiers(final @NotNull String subscriber, final int qos
         this.qos = qos;
         this.flags = flags;
         this.sharedName = sharedName;
-        this.subscriptionIdentifier = ImmutableIntArray.of();
+        this.subscriptionIdentifier = ImmutableIntArray.of(subscriptionIdentifier);
```
Comment (schaebo) [inline src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java:59]:
```suggestion
        this.subscriptionIdentifier = (subscriptionIdentifier == null) ? ImmutableIntArray.of():ImmutableIntArray.copyOf(subscriptionIdentifier);
```

**Review by schaebo — CHANGES_REQUESTED** [review]

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java (hunk 1 above)
Comment (schaebo) [reply src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java:59]: Thanks for your contribution. 
When creating an ImmutableIntArray from a Collection another signature must be used.
This constructor is only used in a few unit tests that do not test subscription identifiers as such.

Comment (sakib-malek) [reply src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java:59]: Yes, your suggestion makes sense.

However, given the `NotNull` annotations, should not the **nullability** be check by `checkNotNull`?

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java [hunk 2]
```diff
@@ -52,11 +52,12 @@ public SubscriberWithIdentifiers(final @NotNull String subscriber, final int qos
     public SubscriberWithIdentifiers(final @NotNull String subscriber, final int qos, final byte flags, final @Nullable String sharedName,
                                      final @NotNull ImmutableList<Integer> subscriptionIdentifier, final @Nullable String topicFilter) {
         checkNotNull(subscriber, "Subscriber must not be null");
+        checkNotNull(subscriptionIdentifier, "Subscriber Identifiers must not be null")
```
Comment (schaebo) [inline src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java:55]:
```suggestion
        checkNotNull(subscriptionIdentifier, "Subscriber Identifiers must not be null");
```

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java [hunk 3]
```diff
@@ -52,11 +52,12 @@ public SubscriberWithIdentifiers(final @NotNull String subscriber, final int qos
...
         this.subscriber = subscriber;
         this.qos = qos;
         this.flags = flags;
         this.sharedName = sharedName;
-        this.subscriptionIdentifier = ImmutableIntArray.of();
+        this.subscriptionIdentifier = (subscriptionIdentifier == null) ? ImmutableIntArray.of():ImmutableIntArray.copyOf(subscriptionIdentifier);
```
Comment (schaebo) [inline src/main/java/com/hivemq/mqtt/topic/SubscriberWithIdentifiers.java:60]: With the added `checkNotNull` the null check here is, as you already mentioned, not necessary anymore and can be removed.

**Review by schaebo — CHANGES_REQUESTED** [review]

**Comment by schaebo** [comment]
@sakib-malek: Thank you for incorporating the changes. Unit tests are currently failing for this change.

Excerpt from travis build:

```
com.hivemq.mqtt.InternalPublishServiceImplTest > test_shared_subs_same_group FAILED
    java.lang.NullPointerException: Subscriber Identifiers must not be null
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:897)
        at com.hivemq.mqtt.topic.SubscriberWithIdentifiers.<init>(SubscriberWithIdentifiers.java:55)
        at com.hivemq.mqtt.InternalPublishServiceImplTest.test_shared_subs_same_group(InternalPublishServiceImplTest.java:244)
com.hivemq.mqtt.InternalPublishServiceImplTest > test_mixed_subs FAILED
    java.lang.NullPointerException: Subscriber Identifiers must not be null
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:897)
        at com.hivemq.mqtt.topic.SubscriberWithIdentifiers.<init>(SubscriberWithIdentifiers.java:55)
        at com.hivemq.mqtt.InternalPublishServiceImplTest.test_mixed_subs(InternalPublishServiceImplTest.java:267)
com.hivemq.mqtt.InternalPublishServiceImplTest > test_shared_subs_different_groups FAILED
    java.lang.NullPointerException: Subscriber Identifiers must not be null
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:897)
        at com.hivemq.mqtt.topic.SubscriberWithIdentifiers.<init>(SubscriberWithIdentifiers.java:55)
        at com.hivemq.mqtt.InternalPublishServiceImplTest.test_shared_subs_different_groups(InternalPublishServiceImplTest.java:221)

```

Can you fix them ?
If you need help with that, let me know.

**Comment by DC2-DanielKrueger** [comment]
I closed the PR, because no progress was made within the last 6 months and it has test failures.

Best regards, 

Daniel

**Pull request closed without merging**